protocol/contracts/beanstalk/sun/SeasonFacet/Weather.sol

```diff
@@ -193,111 +311,96 @@ contract Weather is Sun {
-        (uint256[] memory ratios, uint256 beanIndex, bool success) = LibWell.getRatiosAndBeanIndex(
-            tokens
-        );
-        // If the USD Oracle oracle call fails, the sop should not occur.
-        // return 0 rather than revert to prevent sunrise from failing.
-        if (!success) return (0, IERC20(address(0)));
-
-        // compare the beans at peg using the instantaneous reserves,
-        // and the current reserves.
-        uint256 instantaneousBeansAtPeg = IBeanstalkWellFunction(wellFunction.target)
-            .calcReserveAtRatioSwap(instantaneousReserves, beanIndex, ratios, wellFunction.data);
-
-        uint256 currentBeansAtPeg = IBeanstalkWellFunction(wellFunction.target)
-            .calcReserveAtRatioSwap(currentReserves, beanIndex, ratios, wellFunction.data);
+    // code review note: have some casual comments here for just understanding the code, happy to
+    // change this function to make it more efficient based on feedback.
+    function calculateSopPerWell(
+        WellDeltaB[] memory wellDeltaBs,
+        uint256 totalPositiveDeltaB,
+        uint256 totalNegativeDeltaB,
+        uint256 positiveDeltaBCount
+    ) public pure returns (WellDeltaB[] memory) {
+        // most likely case is that all deltaBs are positive
+        if (positiveDeltaBCount == wellDeltaBs.length) {
+            // if all deltaBs are positive, need to sop all to zero, so return existing deltaBs
+            return wellDeltaBs;
+        }
 
-        // Calculate the signed Sop beans for the two reserves.
-        int256 lowestSopBeans = int256(instantaneousBeansAtPeg).sub(
-            int256(instantaneousReserves[beanIndex])
-        );
-        int256 currentSopBeans = int256(currentBeansAtPeg).sub(int256(currentReserves[beanIndex]));
+        if (positiveDeltaBCount == 0) {
+            // No positive values, so no well flooding needed, return zeros
+            for (uint256 i = 0; i < positiveDeltaBCount; i++) {
+                wellDeltaBs[i].deltaB = 0;
+            }
+            return wellDeltaBs;
+        }
 
-        // Use the minimum of the two.
-        if (lowestSopBeans > currentSopBeans) {
-            lowestSopBeans = currentSopBeans;
+        if (totalPositiveDeltaB < totalNegativeDeltaB) {
+            // This can occur if the twaDeltaB is positive, but the instanteous deltaB is negative or 0
+            // In this case, no reductions are needed.
+            for (uint256 i = 0; i < positiveDeltaBCount; i++) {
+                wellDeltaBs[i].deltaB = 0;
+            }
+            return wellDeltaBs;
         }
 
-        // If the sopBeans is negative, the sop should not occur.
-        if (lowestSopBeans < 0) return (0, IERC20(address(0)));
+        uint256 shaveToLevel = totalNegativeDeltaB / positiveDeltaBCount;
 
-        // SafeCast not necessary due to above check.
-        sopBeans = uint256(lowestSopBeans);
+        for (uint256 i = positiveDeltaBCount; i > 0; i--) {
```

Comment:
can we have a comment here explaining what the for loop does?

Reply:
Added here: https://github.com/BeanstalkFarms/Beanstalk/pull/870/commits/6b55d95bc5a8d90f6bafddfe6e78c45ee7be0e0f